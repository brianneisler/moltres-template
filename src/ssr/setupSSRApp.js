import * as modules from '../app/modules'
import { App, generateConfig } from '../app'
import { AppRegistry } from 'react-native-web'
import {
  cacheMethod,
  isArray,
  isFunction,
  isImmutable,
  isObject,
  omit,
  walkMap
} from '../utils/data'
import { createEngine, selectUncaughtException, setup, start } from '../core'
import { createHistory } from '../utils/react'
import { locationChangeAction } from '../app/modules/router/actions'
import { parseURL } from '../utils/url'
import { selectRouterResponse } from '../app/modules/router'
import MetaTagsServer from 'react-meta-tags/server'
import ReactDOMServer from 'react-dom/server'
import setupSSRContext from './setupSSRContext'

// This code is based on SSR for react-native-web
// https://github.com/necolas/react-native-web/blob/master/docs/guides/server-side-rendering.md

const filterState = (state) => {
  // NOTE BRN: Filter the state for any values that we don't want to
  // send to the client side. Also filter any values that are complex objects.
  return {
    query: walkMap((value) => {
      if (isImmutable(value)) {
        value = value.toJSON()
      }
      if (isObject(value) && !isArray(value) && !isFunction(value)) {
        return omit(['nextPage', 'watch', 'cursor', 'cursorNext'], value)
      }
      return value
    }, state.query)
  }
}

const waitForResponse = async (store) => {
  let response = selectRouterResponse(store.getState())
  if (response) {
    return response
  }
  return new Promise((resolve) => {
    store.subscribe(() => {
      response = selectRouterResponse(store.getState())
      if (response) {
        return resolve(response)
      }
    })
  })
}

const monitorForError = async (store) => {
  let exception = selectUncaughtException(store.getState())
  if (exception) {
    throw exception
  }
  return new Promise((resolve, reject) => {
    store.subscribe(() => {
      exception = selectUncaughtException(store.getState())
      if (exception) {
        return reject(exception)
      }
    })
  })
}

const setupSSRApp = () => {
  let renderer
  const setupRenderer = () => {
    AppRegistry.registerComponent('App', () => App)
    return {
      render: async (context, url) => {
        const config = generateConfig({
          ssr: true,
          stage: context.config.stage,
          test: context.config.test
        })
        const { cache } = context
        const metaTagsInstance = MetaTagsServer()
        const history = createHistory(url)
        const ssrContext = setupSSRContext(config, context, history)
        const store = start(setup(createEngine(modules, config, ssrContext)))
        const initialProps = {
          cache,
          history,
          metaTagsInstance,
          store
        }
        const { element, getStyleElement } = AppRegistry.getApplication('App', {
          initialProps
        })
        const location = parseURL(url)

        await store.dispatch(
          locationChangeAction(context, {
            location
          })
        )
        const html = ReactDOMServer.renderToString(element)
        const response = await Promise.race([
          waitForResponse(store),
          monitorForError(store)
        ])
        const state = filterState(store.getState())

        await store.stop(store)
        return {
          css: ReactDOMServer.renderToStaticMarkup(getStyleElement({})),
          html,
          meta: metaTagsInstance.renderToString(),
          response,
          state
        }
      }
    }
  }

  const cachedRenderSSR = cacheMethod(
    {
      // TODO BRN: If we start rendering pages on a per user basis we'll need to
      // use the context.currentUser as part of the key. Need to figure out a
      // way to allow for some pages to be cached
      key: (context, request) => ['page', request.url],
      onHit: ({ logger }, request) =>
        logger.info(`Cache hit for request for ${request.url}`),
      ttl: 60 * 60 * 1000
    },
    async (context, request) => {
      const { logger } = context
      logger.info(`Cache miss for request for ${request.url}`)
      if (!renderer) {
        logger.info(`No renderer found. Building React SSR renderer...`)
        renderer = setupRenderer()
      }

      // TODO BRN: This does not take into account things like pages that should be
      // rendered differently for different users.

      // NOTE BRN: We use request.url here instead of request.originalUrl
      // because in local development the original.url is the path of the
      // firebase app and function which is not the url our app wants

      // NOTE BRN: This is where to feed in things like the current path based upon
      // the http request, the current user, etc
      return await renderer.render(context, request.url)
    }
  )

  return async (request) => {
    const { context } = request
    const { logger } = context
    logger.info(`Rendering request for ${request.url}`)
    return await cachedRenderSSR(context, request)
  }
}

export default setupSSRApp
