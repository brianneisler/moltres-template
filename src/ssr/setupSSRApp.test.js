/**
 * @jest-environment node
 */

import {
  setupTestAdminContext,
  setupTestServiceAccountContext,
  tearDownTestAdminContext,
  tearDownTestServiceAccountContext
} from 'moltres/test'

import setupSSRApp from './setupSSRApp'

const spec = describe('setupSSRApp', () => {
  describe('ServiceAccount', () => {
    let adminContext
    let context

    beforeEach(async () => {
      adminContext = await setupTestAdminContext(spec)
      context = await setupTestServiceAccountContext(adminContext)
    }, 20000)

    afterEach(async () => {
      context = await tearDownTestServiceAccountContext(context)
      adminContext = await tearDownTestAdminContext(adminContext)
    }, 20000)

    it('renders SSR app without error', async () => {
      const render = setupSSRApp()

      expect(typeof render).toBe('function')

      const request = {
        context,
        url: '/'
      }
      const result = await render(request)

      expect(result).toMatchSnapshot()
    }, 20000)
  })
})
