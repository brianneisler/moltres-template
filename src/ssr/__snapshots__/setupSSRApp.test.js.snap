// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupSSRApp ServiceAccount renders SSR app without error 1`] = `
Object {
  "css": "<style id=\\"react-native-stylesheet\\">[stylesheet-group=\\"0\\"]{}
html{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);}
body{margin:0;}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0;}
input::-webkit-inner-spin-button,input::-webkit-outer-spin-button,input::-webkit-search-cancel-button,input::-webkit-search-decoration,input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none;}
[stylesheet-group=\\"0.1\\"]{}
:focus:not([data-focusvisible-polyfill]){outline: none;}
[stylesheet-group=\\"1\\"]{}
.css-view-1dbjc4n{-ms-flex-align:stretch;-ms-flex-direction:column;-ms-flex-negative:0;-ms-flex-preferred-size:auto;-webkit-align-items:stretch;-webkit-box-align:stretch;-webkit-box-direction:normal;-webkit-box-orient:vertical;-webkit-flex-basis:auto;-webkit-flex-direction:column;-webkit-flex-shrink:0;align-items:stretch;border:0 solid black;box-sizing:border-box;display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;flex-basis:auto;flex-direction:column;flex-shrink:0;margin-bottom:0px;margin-left:0px;margin-right:0px;margin-top:0px;min-height:0px;min-width:0px;padding-bottom:0px;padding-left:0px;padding-right:0px;padding-top:0px;position:relative;z-index:0;}
.css-text-901oao{border:0 solid black;box-sizing:border-box;color:rgba(0,0,0,1.00);display:inline;font:14px system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,Ubuntu,\\"Helvetica Neue\\",sans-serif;margin-bottom:0px;margin-left:0px;margin-right:0px;margin-top:0px;padding-bottom:0px;padding-left:0px;padding-right:0px;padding-top:0px;white-space:pre-wrap;word-wrap:break-word;}
[stylesheet-group=\\"2.2\\"]{}
.r-pointerEvents-12vffkv>*{pointer-events:auto;}
.r-pointerEvents-12vffkv{pointer-events:none!important;}</style>",
  "html": "<div class=\\"css-view-1dbjc4n r-pointerEvents-12vffkv\\" style=\\"-webkit-flex-grow:1;flex-grow:1;-webkit-flex-shrink:1;flex-shrink:1;-webkit-flex-basis:0%;flex-basis:0%;-ms-flex-positive:1;-webkit-box-flex:1;-ms-flex-negative:1;-ms-flex-preferred-size:0%\\" data-reactroot=\\"\\"><div class=\\"css-view-1dbjc4n r-pointerEvents-12vffkv\\" style=\\"-webkit-flex-grow:1;flex-grow:1;-webkit-flex-shrink:1;flex-shrink:1;-webkit-flex-basis:0%;flex-basis:0%;-ms-flex-positive:1;-webkit-box-flex:1;-ms-flex-negative:1;-ms-flex-preferred-size:0%\\"><div class=\\"css-view-1dbjc4n\\" style=\\"-webkit-align-items:center;align-items:center;background-color:rgba(255,255,255,1.00);-webkit-flex-grow:1;flex-grow:1;-webkit-flex-shrink:1;flex-shrink:1;-webkit-flex-basis:0%;flex-basis:0%;-webkit-justify-content:center;justify-content:center;-ms-flex-align:center;-webkit-box-align:center;-ms-flex-positive:1;-webkit-box-flex:1;-ms-flex-negative:1;-ms-flex-preferred-size:0%;-ms-flex-pack:center;-webkit-box-pack:center\\"><div dir=\\"auto\\" class=\\"css-text-901oao\\">Open up App.js to start working on your app!</div></div></div></div>",
  "meta": "",
  "response": Object {
    "error": [Error: This page does not exist],
    "statusCode": 404,
  },
  "state": Object {
    "query": Object {},
  },
}
`;
